source/common/unified_memory_allocation/src/memory_pool.c

```diff
@@ -6,42 +6,61 @@
  *
  */
 
+#include "memory_tracker_internal.h"
 #include <uma/memory_pool.h>
 
 #include <assert.h>
 #include <stdlib.h>
 
 struct uma_memory_pool_t {
+    void *pool_priv; // so that &pool->pool_priv == pool
     struct uma_memory_pool_ops_t ops;
-    void *pool_priv;
+    uma_memory_provider_handle_t *providers;
+    size_t numProviders;
 };
 
-enum uma_result_t umaPoolCreate(struct uma_memory_pool_ops_t *ops, void *params,
-                                uma_memory_pool_handle_t *hPool) {
+enum uma_result_t umaPoolCreate(struct uma_memory_pool_ops_t *ops, uma_memory_provider_handle_t *providers,
+                                size_t numProviders, void *params, uma_memory_pool_handle_t *hPool) {
     uma_memory_pool_handle_t pool = malloc(sizeof(struct uma_memory_pool_t));
     if (!pool) {
         return UMA_RESULT_ERROR_OUT_OF_HOST_MEMORY;
     }
 
     assert(ops->version == UMA_VERSION_CURRENT);
 
-    pool->ops = *ops;
+    pool->providers = calloc(numProviders, sizeof(uma_memory_provider_handle_t));
+    if (!pool->providers) {
+      free(pool);
+      return UMA_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+  
+    pool->numProviders = numProviders;
+    for (size_t i = 0; i < numProviders; i++) {
+      enum uma_result_t ret = umaTrackingMemoryProviderCreate(providers[i], (uma_memory_pool_handle_t)pool, &pool->providers[i]);
```

Comment:
I have another concern: technically the same provider instance might be used by multiple pools. At least our API does not prohibit this.
Strictly speaking, to determine the pool by ptr we need to map ptr to heap manager that requested allocation. But for the MPI use case we need to determine the properties of the memory - it means that we need to map ptr to the memory provider. Because the memory provider defines the properties of the memory.